### PR TITLE
Fix TS type issues w/ the packages, add ES2015 UMD bundles

### LINKED ISSUE #28: Modules Not Found in Angular CLI Application
Within an Angular/CLI application (in VS Code), ES imports cannot resolve the modules.  For example, these imports fail...

```
import { IterableX as Iterable } from 'ix/iterable';
import 'ix/add/iterable/of';
import 'ix/add/iterable-operators/map';
import { map } from 'ix/iterable/map';
import { filter } from 'ix/iterable/filter';
```

However, this import does work...

```
import * as Ix from 'ix';
```
That, sadly, does not give me access to 'ix/iterable' functions.

### PR DESCRIPTION
**Pull request by trxcllnt**
This PR should fix #28, #30, and #43 once it's merged and published

We've added a separate build target for the main `ix` module whose directory structure aligns 1-1 with the structure of the source files. The main repository is still going to be larger than the scoped packages, because we're packaging multiple compilation targets side-by-side:

- ES5/UMD bundle is in there for older browsers as `Ix.es5.min.js`
- ES6/UMD bundle is in there for _newer_ browsers as `Ix.es2015.min.js` (new!)
- ES5/CommonJS lowest common denominator are in as `*.js` files
- ES2015/ESModule targets are in as `*.mjs`, for compatibility with node
- TypeScript sourcefiles are in as `*.ts` files. This alleviates the need for us to compile `d.ts` types (intellisense will jump you to the actual sources), and it lets people configure their bundlers to operate on the original TS source w/o much headache (specifically, I think `@angular/cli` does this)

I've validated this works well in VSCode, angular/cli, create-react-app, and vanilla TS projects. It doesn't work with `create-react-app-ts` but it's a webpack configuration thing, and it does work if you change the setup ([more here](https://github.com/wmonk/create-react-app-typescript/issues/168)).

Also, the build is wired up via Observables now, and is quite a bit faster than before 🎉

### CODE REVIEW

File: tsconfig/tsconfig.esnext.cls.json [hunk 1]
```diff
@@ -0,0 +1,12 @@
+// Compiler configuaration to build the intermediate closure-compiler ESM UMD target.
+// Not used at the moment, but ready whenever CC has full ES2015+ support.
```
Comment (trxcllnt) [inline tsconfig/tsconfig.esnext.cls.json]: I just noticed these comments are out of date as of yesterday, these files _are_ being used to create the intermediate `cls` target that gets Ugly-fied

**Comment by mattpodwysocki** [comment]
LGTM

**Comment by felixfbecker** [comment]
> TypeScript sourcefiles are in as *.ts files. This alleviates the need for us to compile d.ts types

This doesn't sound like a good idea. Maybe I'm missing somethign but wouldn't this require the consumer to have the exact same TS compilation settings (`target`, `module`, `lib` etc) as IxJS?

I see the argument for jumping to the source but if the consumer needs to do that often then I would rather improve the documentation (just the docblock and interface declaration in the .d.ts files should give enough information to answer any kind of question a consumer might have about a function).

**Comment by trxcllnt** [comment]
> @felixfbecker wouldn't this require the consumer to have the exact same TS compilation settings (target, module, lib etc) as IxJS?

I didn't encounter these issues while testing `tsc`, `@angular/cli`, and `create-react-app-ts`. With the linked issue fixed, `create-react-app-ts` uses the ESM JS. 

`tsc` and `@angular/cli` do prefer the `ts` over the compiled JS, so the compilation is dependent on the consumer's tsconfig. The only Ix-specific changes I had to make were:
1. add `"esnext.asynciterable"` to libs
2. `"downLevelIteration": true` if targeting es5

Both these configs would be necessary if `tsc` was processing the JS too, since presumably people are using `for..of` and/or `for await..of` in their app. I do think it'll be important to note this in the readme somewhere.

The two main benefits I see are:
1. if someone is targeting es2015 in their tsconfig, they don't end up with downleveled generator functions
2. The Ix TS source is available for intellisense, which I can't stress enough how valuable that is for people trying to learn how Ix works

Also, let me know if I missed anything in my testing!

**Comment by felixfbecker** [comment]
@trxcllnt I don't know much about `@angular/cli`, `create-react-app-ts` etc but publishing only .js and .d.ts in one package is definitely the best practice for every TS library I've seen out there. Every compiled .js file should have a corresponding .d.ts file next to it with the same basename, and if package.json has a `main` field it should have a corresponding `types` field. With that setup I've never had a problem with any TS library. Here are the benefits of that:
- The `compilerOptions` of the library are not coupled to the `compilerOptions` of the consuming project. For example, if IxJS decides to make use of decorators, you can just enable that flag in your `compilerOptions` without requiring the consumer to also add the flag. It is simply transpiled to JS, the consumer won't notice.
- Consumers who don't write TypeScript but JavaScript (and therefor install the generic CommonJS bundle) can import the package and get autocompletion by editors like Visual Studio Code from the definition files that are included.
- I only need to compile my own project, not all the libraries I'm pulling in. If my project has 50 files the compiler should only have to do the work of compiling and type checking these 50 files, not the 200 files I'm bringing in through various dependencies, or the build will become slow eventually and require an increasing amount of memory.

> The Ix TS source is available for intellisense, which I can't stress enough how valuable that is for people trying to learn how Ix works

Totally agree, but IntelliSense works completely fine with .d.ts files. It works even for JavaScript as described above

> if someone is targeting es2015 in their tsconfig, they don't end up with downlevelled generator functions

If this is really a problem, I would do this:
- The module bundle stays close to latest features, lets say always supports the oldest currently active Node LTS version. Atm that would be Node 4 which supports generators. If consumers of this package need to use the package in the browser, since this is a module package they will have a bundler build pipeline like webpack anyway. So if they need to target older browsers it's trivial to bring in `babel-preset-env` and transpile down to the browser they need. This package includes the .d.ts files for TS consumers or JS consumers who want IntelliSense. I am not familiar with .mjs loading but it could be the just one package with .js and .mjs sitting next to each other. 
- The distributed, minfied browser bundle will transpile downlevel as low as possible. The consumers of this package are expected to not use a bundler (otherwise they would use the module build for its various benefits like minimal imports) so they probably don't have a build pipeline. Again, this could be a separate package but for simplicity it could also just be inside a single .min.js file in a dist/ folder like with most packages.

These are just my 2ct and as said please take it with a grain of salt as I have never worked with the Angular CLI and the other projects mentioned. This is just my experience from publishing and consuming TS projects myself.

**Comment by trxcllnt** [comment]
@felixfbecker I appreciate you taking the time to write here, and I have responses to the issues you've raised below, but maybe it'll be quicker to `npm install ix` the release we just published so you can see what I'm talking about on your own time :-)

> The compilerOptions of the library are not coupled to the compilerOptions of the consuming project. For example, if IxJS decides to make use of decorators, you can just enable that flag in your compilerOptions without requiring the consumer to also add the flag

I get what you're saying, but we won't be using decorators. To import and use Ix, someone has to add "esnext.asyncIterable" in their tsconfig, regardless of whether they're using TS's downleveled ES5 output or not:

screen shot 2017-10-02 at 7 07 43 pm
screen shot 2017-10-02 at 7 06 26 pm

> Every compiled .js file should have a corresponding .d.ts file next to it with the same basename, and if package.json has a main field it should have a corresponding types field.

> IntelliSense works completely fine with .d.ts files. It works even for JavaScript as described above

The issue is .d.ts files are only the type signatures, not the source code. Type signatures already mean very little in JavaScript, and are next to worthless for learning the details of how interesting projects like Ix and Rx actually work.

> - Consumers who don't write TypeScript but JavaScript (and therefor install the generic CommonJS bundle) can import the package and get autocompletion by editors like Visual Studio Code from the definition files that are included.

Yes, this is also the case when the original `*.ts` sources are published instead of the dtypes. VSCode and `tsc` are equally happy when the "types" field in package.json refers to a '.ts' source file instead of a dtype.

> - I only need to compile my own project, not all the libraries I'm pulling in. If my project has 50 files the compiler should only have to do the work of compiling and type checking these 50 files, not the 200 files I'm bringing in through various dependencies, or the build will become slow eventually and require an increasing amount of memory.

This is a reasonable thing to expect. Most JS build tools will ignore `ts` files in node_modules, or can be trivially configured to do so. Also, we _do_ publish the compiled JS in the `ix` package (or rather, multiple versions of the compiled JS), so we can support as many module formats/JS environments going forward.

> > if someone is targeting es2015 in their tsconfig, they don't end up with downlevelled generator functions

> If this is really a problem

It actually is a huge problem in projects like [Apache Arrow](https://github.com/apache/arrow/tree/master/js), where the downleveled code is significantly slower than native iterators. All the generated code also contributes ~26k to the ES5 UMD bundle vs. the ES6 one (not to mention the sourcemap sizes), so for people who want to [deploy es2015+ modules](https://philipwalton.com/articles/deploying-es2015-code-in-production-today/) for the newer browsers today, we support that:

screen shot 2017-10-02 at 7 46 29 pm

> I would do this:
> 
> - The module bundle stays close to latest features, lets say always supports the oldest currently active Node LTS version. Atm that would be Node 4 which supports generators. If consumers of this package need to use the package in the browser, since this is a module package they will have a bundler build pipeline like webpack anyway. So if they need to target older browsers it's trivial to bring in `babel-preset-env` and transpile down to the browser they need.

That's my opinion as well, but if configuring build tools were so easy, we wouldn't have needed to rethink our publishing strategy to begin with ;). But seriously, this is about finding a strategy for the main published `ix` module that covers 80% of the use-cases out there (ES5 CommonJS, ES5 UMD, es2015 ESModules, es2015 UMD, `<script type="module">`, node, newer but still old node, nightlies of node, etc.).

> - This package includes the .d.ts files for TS consumers or JS consumers who want IntelliSense. I am not familiar with .mjs loading but it could be the just one package with .js and .mjs sitting next to each other.

`.mjs` is pretty broken right now (and has zero TypeScript support), but what we have now does work with how it's implemented in node 8.6.0.

> - The distributed, minfied browser bundle will transpile downlevel as low as possible. The consumers of this package are expected to not use a bundler (otherwise they would use the module build for its various benefits like minimal imports) so they probably don't have a build pipeline. Again, this could be a separate package but for simplicity it could also just be inside a single .min.js file in a dist/ folder like with most packages.

Take a look at the files in `ix@2.1.2` on npm. It's pretty much what you're describing, except `.ts` files instead of `.d.ts`, and includes the ES5 and ES2015 minified forms:

screen shot 2017-10-02 at 7 33 48 pm

We also still publish each target in [separate scoped packages](https://github.com/ReactiveX/IxJS#packaging) for anyone who can't afford 7.8 mb of transpiled Ix sitting in their node_modules folder.

**Pull request merged**